### ISSUE TITLE
PATCH public issue with accepted content and title

### ISSUE THREAD

**Issue by miketaylr** [webcompat/webcompat.com#3140]
If an issue is moderated as acceptable we need to do the following (regardless of how or where we stored bug reports from the previous section):

1. Using the GitHub API, we send a PATCH request to the existing public issue and replace the body with the stored private details.
2. Make sure the labels match. Probably good to add a “status-moderation-accepted” label until we decide it's useless.

**Comment by miketaylr** [comment]
Depends on #3150.

**Comment by karlcow** [comment]
@ksy36 @miketaylr This is an RFC.

In fact there is no dependency on #3150. 
Probably #3141 is exactly the same than this here.

But there are probably two missing  milestones: 

* `accepted`
* `not-acceptable`

Let's think this

When the issue on `PRIVATE_REPO_URI` is assigned the milestone `accepted`, then a webhook will start the publishing process. 

1. monitor milestone changes
2. When `milestone == 'accepted'`
3. Read the issue content on `PRIVATE_REPO_URI`. The full content should already be formatted. 
4. With `webcompat-bot`, post to `ISSUES_REPO_URI` the body with an additional `status-moderation-accepted`.
5. Close (and lock?) the issue on `PRIVATE_REPO_URI`

This might work.

Where I'm a bit worried is the  `not-acceptable` milestone and human errors, which has the same mechanism but instead post a template saying no. 


**Let's factor in, the human error.** 


1. if it's a good issue, selecting by mistake `not acceptable`. It will do no harm. The person just need to reopen the issue and change to accepted.
2. if it's a bad issue, selecting by mistake `accepted` has bad consequences.

Probably we should avoid the `not acceptable` milestone and use another trigger for closing issues. To `close a bad issue`, maybe we should just do that **close the issue**. Then the webhook will monitor the status change and will patch the PUBLIC REPO to close the issue.

**Comment by miketaylr** [comment]
>  Probably #3141 is exactly the same than this here.

Yeah, likely they can both be closed at the same time.

> When the issue on PRIVATE_REPO_URI is assigned the milestone accepted, then a webhook will start the publishing process.

I was thinking about this a little differently, but also... pretty close to your idea. 

The idea is basically captured in #3144 (which can also be closed as a dupe of this bug). My thinking was that the private and public repos have the same milestones, with the exception of an additional `unmoderated` milestone in the private repo.

One of 3 things is then possible:

1. Someone does pure moderation: they moderate the issue, find it acceptable, and move it to the "needstriage" milestone.
2. Someone does moderation, they moderate the issue, find it unacceptable, and move it to the "wontfix" milestone (edit: I see this is ambiguous with point 3 below, so probably an "unacceptable" milestone is useful here.
3. Someone does moderation + triage at the same time, finds it acceptable, and moves it to the appropriate milestone (needsdiagnosis, contactready, duplicate, invalid, etc.).

In the case of 1, we would have the webhook notice that it went to needstriage, so it would find the matching public issue already in needstriage and remove the "in moderation" label (or whatever we decide upon), and clone over the issue contents, labels, etc. Do the PATCH dance.

In the case of 2, the webhook notices it went to "unacceptable" and so it updates the public issue contents with the moderation failed template, closes it, and probably sticks it in the WONTFIX milestone (so we don't pollute invalid milestone which ML bots are trained on).

In the case of 3, the webhook would notice it went to a milestone other than "unacceptable" (or wasn't closed, as you propose), and so it clones the issue, and moves the public issue to the corresponding milestone in the public repo, and matches the issue state (for invalid, duplicate, etc.).

> To `close a bad issue`, maybe we should just do that **close the issue**

My one concern with this idea is that I accidentally close issues way more often than I accidentally change milestones. How do we undo the "closed as unacceptable" mistake? Re-open?

**Comment by karlcow** [comment]
> My one concern with this idea is that I accidentally close issues way more often than I accidentally change milestones. How do we undo the "closed as unacceptable" mistake? Re-open?

yes and nothing has been leaked!

While an error in selecting the wrong milestones will leak things.

**Comment by miketaylr** [comment]
OK! let's go for it and see how it "feels". We can move to something different if it doesn't work out in practice.

**Comment by karlcow** [comment]
#3140 #141 and #3155 will be dealt in the same pull request probably.

**Comment by karlcow** [comment]
Notes to self:

* We need to make sure about the origin of the request for the `opened` action. So we do not spend too much time on useless or counter-effective actions. (aka ignore `opened` on the private repo).
* We need to watch the `milestoned` with value `accepted` on the private repo to trigger the action.
* We need to watch the `closed` action (private repo) to trigger the patching with moderation status. Think about reusing the moderation template function.
* We need to separate in clear functions  to make it a bit easier to refactor. 
* When it has been moderated (accepted or refused). We probably need to remove the `action-needsmoderation` label.

**Comment by karlcow** [comment]
These data could be useful. 

*  action: milestoned.
* issue.milestone.title: accepted
* issue.milestone.updated_at: 2020-01-22T01:02:03Z
* issue.body: 'long string'

the event edited keeping interesting information.

```json
{
  "action": "milestoned",
  "issue": {
    "url": "https://api.github.com/repos/webcompat/web-bugs/issues/47787",
    "repository_url": "https://api.github.com/repos/webcompat/web-bugs",
    "number": 47787,
    "title": "vimeo.com - see bug description",
    "milestone": {
      "url": "https://api.github.com/repos/webcompat/web-bugs/milestones/2",
      "html_url": "https://github.com/webcompat/web-bugs/milestone/2",
      "number": 2,
      "title": "needstriage",
      "updated_at": "2020-01-22T01:02:03Z",
    },
    "body": "…the usual webcompat body…"
  }
}
```
